File: src/Modules/Notifications/Lister.Notifications.Domain/Views/NotificationRule.cs
using System.Text.Json.Serialization;

namespace Lister.Notifications.Domain.Views;

public class NotificationRule : IReadOnlyNotificationRule
{
    public object Trigger { get; set; } = null!;
    public object[] Channels { get; set; } = null!;
    public object Schedule { get; set; } = null!;
    public bool IsActive { get; set; }
    public string? TemplateId { get; set; }

    [JsonPropertyName("id")]
    public Guid? Id { get; set; }

    [JsonPropertyName("userId")]
    public string UserId { get; set; } = null!;

    [JsonPropertyName("listId")]
    public Guid ListId { get; set; }
}

File: src/Lister.Client/src/models/notification-rule.ts
export type NotificationTriggerType =
  | "ItemCreated"
  | "ItemDeleted"
  | "ItemUpdated"
  | "StatusChanged"
  | "ColumnValueChanged"
  | "ListDeleted"
  | "ListUpdated"
  | "CustomCondition";

export interface NotificationTrigger {
  type: NotificationTriggerType;
  fromValue?: string | null;
  toValue?: string | null;
  columnName?: string | null;
  operator?: string | null;
  value?: string | null;
}

export type NotificationChannelType =
  | "Email"
  | "Sms"
  | "InApp"
  | "Push"
  | "Webhook";

export interface NotificationChannel {
  type: NotificationChannelType;
  address?: string | null;
  settings?: Record<string, string>;
}

export type NotificationScheduleType =
  | "Immediate"
  | "Delayed"
  | "Daily"
  | "Weekly"
  | "Batched"
  | "Custom";

export interface NotificationSchedule {
  type: NotificationScheduleType;
  delay?: string | null;
  cronExpression?: string | null;
  dailyAt?: string | null;
  daysOfWeek?: string[] | null;
}

export interface NotificationRule {
  id?: string;
  listId: string;
  userId: string;
  trigger: NotificationTrigger;
  channels: NotificationChannel[];
  schedule: NotificationSchedule;
  templateId?: string | null;
  isActive?: boolean;
}

export interface NotificationRuleInput {
  trigger: NotificationTrigger;
  channels: NotificationChannel[];
  schedule: NotificationSchedule;
  templateId?: string | null;
}
